### PR TITLE
S3 based Hive query runner improvements

### PR DESCRIPTION
**Pull request by losipiuk**

### CODE REVIEW

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java [hunk 1]
```diff
@@ -126,6 +127,12 @@ public Builder setInitialTables(Iterable<TpchTable<?>> initialTables)
             return this;
         }
 
+        public Builder setSchemasLocationBase(String schemasLocationBase)
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java]: This is applied only to tpch,tpch_bucketed schema, so name this properly.
(or make it applied universally to all CREATE SCHEMA, which would be even better!)

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java [hunk 2]
```diff
@@ -109,40 +109,49 @@ protected Builder()
...
-            return this;
+            return (B) this;
         }
 
-        public Builder setHiveProperties(Map<String, String> hiveProperties)
+        public B setHiveProperties(Map<String, String> hiveProperties)
         {
             this.hiveProperties = ImmutableMap.copyOf(requireNonNull(hiveProperties, "hiveProperties is null"));
-            return this;
+            return (B) this;
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java:121]: i guess this generates a warning in the IDE, right?

would be good to have 
```
protected B self() { 
return this; 
}
```

with some proper annotations to suppress the warning

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java [hunk 3]
```diff
@@ -109,40 +109,49 @@ protected Builder()
...
         }
 
-        public Builder setInitialTables(Iterable<TpchTable<?>> initialTables)
+        public B addHiveProperty(String key, String value)
+        {
+            this.hiveProperties = ImmutableMap.<String, String>builder()
+                    .putAll(hiveProperties)
+                    .put(key, value)
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java:128]: if we don't envision overrides, make `hiveProperties` an `ImmutableMap.Builder`

else, go via `HashMap` here.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java [hunk 4]
```diff
@@ -44,46 +44,115 @@ public static DistributedQueryRunner create(
...
+    public static class Builder
+            extends HiveQueryRunner.Builder<Builder>
+    {
+        private HostAndPort hiveEndpoint;
+        private HostAndPort s3Endpoint;
+        private String s3AccessKey;
+        private String s3SecretKey;
+        private String bucketName;
+
+        public Builder setHiveEndpoint(HostAndPort hiveEndpoint)
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java:95]: What is "hive endpoint"?

is it metastoreUri or something?

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java [hunk 5]
```diff
@@ -80,13 +80,13 @@ public static DistributedQueryRunner create()
...
-    public static Builder builder()
+    public static Builder<Builder<?>> builder()
     {
-        return new Builder();
+        return new Builder<>();
     }
 
-    public static class Builder
-            extends DistributedQueryRunner.Builder<Builder>
+    public static class Builder<B extends Builder<?>>
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java:88]: I'd prefer a more descriptive generic param name, like `Self`.

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java (hunk 1 above)
Comment (losipiuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java]: I will name it `initialSchemasLocationBase` and keep it for tpch/tpch_bucketed for now. Can be improved later.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java (hunk 5 above)
Comment (losipiuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/HiveQueryRunner.java:88]: renamed to `SELF` - by convention generics params are SELF.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java (hunk 4 above)
Comment (losipiuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java:95]: renamed

**Comment by losipiuk** [comment]
AC

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java [hunk 6]
```diff
@@ -44,46 +44,115 @@ public static DistributedQueryRunner create(
...
+        public DistributedQueryRunner build()
+                throws Exception
+        {
+            requireNonNull(hiveMetastoreEndpoint, "hiveMetastoreEndpoint is null");
+            requireNonNull(s3Endpoint, "s3Endpoint is null");
+            requireNonNull(s3AccessKey, "s3AccessKey is null");
+            requireNonNull(s3SecretKey, "s3SecretKey is null");
+            requireNonNull(bucketName, "bucketName is null");
+
+            addHiveProperty("hive.s3.endpoint", "http://" + s3Endpoint);
```
Comment (findepi) [inline plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java:135]: Can we require that `s3Endpoint` comes with `http://` prepended? Would be much better than just blindly adding it here

**Review by findepi — APPROVED** [review]

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java (hunk 6 above)
Comment (losipiuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java:135]: s3Enpoint is passed as `HostAndPort` class. We are rendering it to string internally.

Comment (findepi) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java:135]: > s3Enpoint is passed as `HostAndPort` class.

can we do the conversion to string and prepending with http:// as early as possible, so we don't need to deal with this here?

Comment (losipiuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java:135]: We can. Then we should   probably do the same for hiveEndpoint.

Comment (losipiuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java:135]: Changed s3Endpoint. Left hiveMetastoreEndpoint as `HostAndPort` matches usage of the latter better.

File: plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java [hunk 7]
```diff
@@ -44,46 +44,115 @@ public static DistributedQueryRunner create(
...
+                                    new ThriftMetastoreConfig(),
+                                    new HdfsEnvironment(new HiveHdfsConfiguration(
+                                            new HdfsConfigurationInitializer(
+                                                    new HdfsConfig(),
+                                                    ImmutableSet.of()),
+                                            ImmutableSet.of()),
+                                            new HdfsConfig(),
+                                            new NoHdfsAuthentication()),
+                                    false)));
+            setInitialSchemasLocationBase("s3a://" + bucketName);
```
[the remark this answers is not in the document]
Comment (losipiuk) [reply plugin/trino-hive/src/test/java/io/trino/plugin/hive/s3/S3HiveQueryRunner.java]: Aparently I cannot:
```
Error:  io.trino.plugin.hive.TestHive3InsertOverwrite.init  Time elapsed: 87.341 s  <<< FAILURE!
io.trino.spi.TrinoException: Got exception: org.apache.hadoop.fs.UnsupportedFileSystemException No FileSystem for scheme "s3"
	at io.trino.plugin.hive.metastore.thrift.ThriftHiveMetastore.createDatabase(ThriftHiveMetastore.java:1005)
	at io.trino.plugin.hive.metastore.thrift.BridgingHiveMetastore.createDatabase(BridgingHiveMetastore.java:164)
	at io.trino.plugin.hive.HiveQueryRunner$Builder.populateData(HiveQueryRunner.java:204)
	at io.trino.plugin.hive.HiveQueryRunner$Builder.build(HiveQueryRunner.java:190)
	at io.trino.plugin.hive.s3.S3HiveQueryRunner$Builder.build(S3HiveQueryRunner.java:159)
	at io.trino.plugin.hive.s3.S3HiveQueryRunner.create(S3HiveQueryRunner.java:51)
	at io.trino.plugin.hive.BaseTestHiveInsertOverwrite.createQueryRunner(BaseTestHiveInsertOverwrite.java:58)
	at io.trino.testing.AbstractTestQueryFramework.init(AbstractTestQueryFramework.java:107)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.testng.internal.MethodInvocationHelper.invokeMethod(MethodInvocationHelper.java:104)
	at org.testng.internal.Invoker.invokeConfigurationMethod(Invoker.java:515)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:217)
	at org.testng.internal.Invoker.invokeConfigurations(Invoker.java:144)
	at org.testng.internal.TestMethodWorker.invokeBeforeClassMethods(TestMethodWorker.java:169)
	at org.testng.internal.TestMethodWorker.run(TestMethodWorker.java:108)
	at java.base/java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1128)
	at java.base/java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:628)
	at java.base/java.lang.Thread.run(Thread.java:829)
Caused by: MetaException(message:Got exception: org.apache.hadoop.fs.UnsupportedFileSystemException No FileSystem for scheme "s3")
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$create_database_result$create_database_resultStandardScheme.read(ThriftHiveMetastore.java:40003)
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$create_database_result$create_database_resultStandardScheme.read(ThriftHiveMetastore.java:39971)
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$create_database_result.read(ThriftHiveMetastore.java:39905)
	at org.apache.thrift.TServiceClient.receiveBase(TServiceClient.java:86)
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.recv_create_database(ThriftHiveMetastore.java:1122)
	at org.apache.hadoop.hive.metastore.api.ThriftHiveMetastore$Client.create_database(ThriftHiveMetastore.java:1109)
	at io.trino.plugin.hive.metastore.thrift.ThriftHiveMetastoreClient.createDatabase(ThriftHiveMetastoreClient.java:155)
	at io.trino.plugin.hive.metastore.thrift.ThriftHiveMetastore.lambda$createDatabase$48(ThriftHiveMetastore.java:996)
	at io.trino.plugin.hive.metastore.thrift.ThriftMetastoreApiStats.lambda$wrap$0(ThriftMetastoreApiStats.java:42)
	at io.trino.plugin.hive.util.RetryDriver.run(RetryDriver.java:130)
	at io.trino.plugin.hive.metastore.thrift.ThriftHiveMetastore.createDatabase(ThriftHiveMetastore.java:994)
	... 20 more
```

**Pull request merged**